Assets/Scripts/Board/Board.cs

```diff
@@ -216,7 +224,96 @@ private void UpdateBoardUI()
+                                            ( Graph.VertexDirection.South, Graph.VertexDirection.Northeast ), } },
+            { TileDirection.Southeast, new() { ( Graph.VertexDirection.Southeast, Graph.VertexDirection.North ),
+                                            ( Graph.VertexDirection.South, Graph.VertexDirection.Northwest ), } },
+        };
+
+        foreach (TerrainTile tile in terrainTiles)
+        {
+            Vector3 tileWorldPosition = terrainTilemap.CellToWorld(tile.position);
+            Graph.TileNode tileNode = new(tileWorldPosition);
+            Debug.Log("Created TileNode at: " + tileWorldPosition);
+
+            Dictionary<TileDirection, Vector3Int> neighbors = GetNeighbors(tile.position);
+
+            // Set vertices if nieghboring tile already exists
+            foreach (var vertexMapping in vertexMappings)
+            {
+                if (graph.HasTile(neighbors[vertexMapping.Key]))
+                {
+                    Graph.TileNode neighbor = graph.GetTile(neighbors[vertexMapping.Key]);
+                    foreach (var map in vertexMapping.Value)
+                    {
+                        tileNode.AddVertex(map.Item1, neighbor.GetVertex(map.Item2));
+                    }
+                }
+            }
+            // else, create new vertex
+            foreach (Graph.VertexDirection direction in Enum.GetValues(typeof(Graph.VertexDirection)))
+            {
+                if (!tileNode.HasVertex(direction))
+                {
+                    Vector3 vertexWorldPosition = tileWorldPosition + Graph.GetVertexOffsetVector(direction);
+                    Graph.VertexNode vertexNode = new(vertexWorldPosition);
+                    Debug.Log("Created " + direction + " VertexNode at: " + vertexWorldPosition);
+
+                    foreach (Graph.VertexDirection neighborVertexDirection in Graph.GetVertexDirectionNeighbors(direction))
+                    {
+                        if (tileNode.HasVertex(neighborVertexDirection))
+                        {
+                            Graph.VertexNode neighborVertexNode = tileNode.GetVertex(neighborVertexDirection);
+
+                            Vector3 edgeWorldPosition = (vertexNode.GetWorldPosition() + neighborVertexNode.GetWorldPosition()) / 2;
+                            Graph.EdgeNode edgeNode = new(edgeWorldPosition);
+                            Debug.Log("Created EdgeNode at: " + edgeWorldPosition);
+
+                            edgeNode.AddVertex(vertexNode);
+                            edgeNode.AddVertex(neighborVertexNode);
+
+                            vertexNode.AddEdge(edgeNode);
+                            neighborVertexNode.AddEdge(edgeNode);
+
+                            graph.AddEdge(edgeNode);
+                        }
+                    }
+                    tileNode.AddVertex(direction, vertexNode);
+                    graph.AddVertex(vertexNode);
+                }
+            }
```

Comment:
Can this be combined so that you instead just do "for each vertex direction, if graph.HasTile(direction) --> use vertex; else add new" so it isn't having to double loop over all 6 neighbors * 2 vertices per neighbor THEN over 6 vertices, so it only loops over 6 vertices and *potentially* checks 2 vertices for the neighbor if exists?

Reply:
Also how much of this logic can be in the actual Graph class?

Reply:
Moved this logic to Graph.GenerateGraph()
I think combining the logic is a bit tricky, especially in ensuring that all the edges are properly linked.